prac_07/project_management.py

```diff
@@ -0,0 +1,197 @@
+        elif option == 'f':
+            filter_projects(projects)
+        elif option == 'a':
+            projects = add_new_project(projects)
+        elif option == 'u':
+            projects = update_project(projects)
+        else:
+            print("Invalid option. Please try again.")
+        print(MENU)
+        option = input(">>> ").lower()
+    save_changes(projects)
+
+
+def load_projects():
+    """Load the project details."""
+    projects = []
+    in_file = open("projects.txt", 'r')  # load file directly without asking user input
+    # in_file = open(FILENAME, 'r')
+    in_file.readline()
+    for line in in_file:
+        parts = line.strip().split('\t')
+        projected = Project(parts[0], parts[1], parts[2], parts[3], parts[4])
+        projects.append(projected)
+    return projects
+
+
+def save_projects(projects):
+    """Save the project details"""
+    with open(FILENAME, 'w') as out_file:
+        print("Name\tStart Date\tPriority\tCost Estimate\tCompletion Percentage", file=out_file)
+        for project in projects:
+            out_file.write(project.line_format() + '\n')
+
+
+def display_projects(projects):
+    """Display by differing the incompleted and completed project."""
+    sorted_projects = sorted(projects, key=lambda x: x.priority)
+
+    print("Incomplete projects:")
+    incomplete_projects = [f"\t{project}" for project in sorted_projects if not project.is_completed()]
+    print('\n'.join(incomplete_projects))
+
+    print("Completed projects:")
+    completed_projects = [f"\t{project}" for project in sorted_projects if project.is_completed()]
+    print('\n'.join(completed_projects))
+
+
+def filter_projects(projects):
+    """Filter the project by year given."""
+    while True:
+        try:
+            date_string = input("Show projects that start after date (dd/mm/yyyy): ")
+            date_formatted = datetime.datetime.strptime(date_string, "%d/%m/%Y").date()
+            break
+        except:
+            print("Invalid date (check for the date format)")
+
+    filtered_projects = [project for project in projects if datetime.datetime.strptime(project.start_date, "%d/%m/%Y").date() > date_formatted]
+
+    display_projects(filtered_projects)
+
+
+def add_new_project(projects):
+    """Add new project."""
+    print("Let's add a new project")
+    name = input("Name: ")
+
+    while True:
+        try:
+            start_date = input("Start date (dd/mm/yyyy): ")
+            datetime.datetime.strptime(
+                start_date, "%d/%m/%Y").date()
+            break
+        except:
+            print("Invalid date (check for the date format)")
+
+    while True:
+        try:
+            priority = int(input("Priority: "))
+            if priority < 0 or priority > 10:
+                print("Priority must be between 0 to 10")
+            else:
+                break
+        except ValueError:
+            print("Invalid priority")
+
+    while True:
```

Comment:
I think you have a great code but just like we have learnt in class, it would be better if we don't use while Ture. You can try using other methods instead. Just a friendly suggestion tho! ^^

Reply:
But what can I use other than while True?

Reply:
You can use if elif else loop

Reply:
Owww right. But I asked the lecturer and she said it's fine to use while True for error checking. So I guess it should be fine right?

Reply:
Yes, then it should be alright!